### ISSUE TITLE
CryptographicException when calling _interaction.GetAllUserGrantsAsync()

### ISSUE THREAD

**Issue by bNobo** [DuendeSoftware/IdentityServer#1479]
**Which version of Duende IdentityServer are you using?**

6.2.0

**Which version of .NET are you using?**

.NET 6

**Describe the bug**

when calling _interaction.GetAllUserGrantsAsync(), if the store contains a grant associated with a key that does not exist anymore, there is a CryptographicException and the function returns 0 grants. There can be dozens of valid grants, if there is only one that is associated with a key that does not exist anymore, the function returns 0 grants. 

**To Reproduce**

1) Add DataProtection at Startup

```cs
services
    .AddDataProtection()
    .PersistKeysToFileSystem(
        new DirectoryInfo(
            keysPath
        )
    );
```
2) Start the application: a new key is automatically created in the keysPath folder
3) Login a user to create a grant
4) Stop the application and remove the key file in keysPath folder
5) Start the application again: a new key is automatically created in the keysPath folder
6) Login a user to create a grant: the database now contains 2 grants, one is valid and one is invalid because the original key does not exist anymore
7) Call GetAllUserGrantsAsync: no grants returned and the exception appears in console window

**Expected behavior**

GetAllUserGrantsAsync should ignore invalid grants and return valid ones.

**Log output/exception with stacktrace**

The exception is:

```
System.Security.Cryptography.CryptographicException: The key {9246e8a3-ebdd-49df-a7af-cf293e108832} was not found in the key ring. For more information go to http://aka.ms/dataprotectionwarning
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.UnprotectCore(Byte[] protectedData, Boolean allowOperationsOnRevokedKeys, UnprotectStatus& status)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.Unprotect(Byte[] protectedData)
   at Microsoft.AspNetCore.DataProtection.DataProtectionCommonExtensions.Unprotect(IDataProtector protector, String protectedData)
   at Duende.IdentityServer.Stores.Serialization.PersistentGrantSerializer.Deserialize[T](String json) in /_/src/Storage/Stores/Serialization/PersistentGrantSerializer.cs:line 114
   at System.Linq.Utilities.<>c__DisplayClass2_0`3.<CombineSelectors>b__0(TSource x)
   at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
   at Duende.IdentityServer.Services.DefaultPersistedGrantService.Join(IEnumerable`1 first, IEnumerable`1 second) in /_/src/IdentityServer/Services/Default/DefaultPersistedGrantService.cs:line 118
   at Duende.IdentityServer.Services.DefaultPersistedGrantService.GetAllGrantsAsync(String subjectId) in /_/src/IdentityServer/Services/Default/DefaultPersistedGrantService.cs:line 100
```

**Additional context**

**Comment by AndersAbel** [comment]
@bNobo I have reproduced the error and am looking into possible ways to solve this.

Currently the error handling will call `LogError()` with the exception and then return an empty collection of grants.

I think that the error logging can be moved to be done per grant. This would log each grant error individually. Those grants that caused a failure are ignored and the rest returned.

An alternative is to collect a list of deserialization/unprotect failures and log an aggregate exception to avoid spamming the logs with individual errors.

I would prefer an aggregate exception to just write one Error log per call to GetAllUserGrantsAsync. @brockallen @josephdecock what do you think?

**Comment by josephdecock** [comment]
I don't have a strong opinion here, but it probably is simpler to implement with individual errors logged, and a single value might be slightly easier to consume/search for in a log than a list?

**Comment by brockallen** [comment]
@AndersAbel is this something we're targeting our 7.0 release for a fix?

**Comment by brockallen** [comment]
PR merged.

**Comment by brockallen** [comment]
@bNobo we released patch 6.3.8 for this today, so please give it a try. thanks

**Comment by bNobo** [comment]
I gave it a try and the patch fixed the issue. Thx !
